### ISSUE TITLE
Tax rate + 100% discount results in negative grand total

### ISSUE THREAD

**Issue by dankocherga** [magento/magento2#10790]
Setting up tax a rule and applying 100% discount in the shopping cart results in a negative grand total.

### Preconditions
1. Magento 2.1.7

### Steps to reproduce
**Admin setup**
1. Create product A with price = 2 762,71
2. Create product B with price = 2 542,37
3. Setup tax rule 18%
4. Create coupon code that adds 100% discount.

**Front-End**
1. Add both products to the cart with qty = 2
2. Apply coupon code

### Expected result
1. Grand total (without shipping yet calculated) = 0

### Actual result
1. Grand total is -0,01

**Calculating by hand**

Product | Qty | Price | Row total | Row Tax
--- | --- | --- | --- | ---
A | 2 | 2762,71 | 5525,42 | 994,5756
B | 2 | 2542,37 | 5084,74 | 915,2532

Total | Expected Amount
--- | ---
subtotal | 10610,16
tax | 1909,8288

**How Magento calculates it** 
Putting a debugger at [Magento\Quote\Model\Quote\Address\Total\Grand](https://github.com/magento/magento2/blob/develop/app/code/Magento/Quote/Model/Quote/Address/Total/Grand.php#L27) we could see:

<img width="456" alt="2017-09-06 11 35 27" src="https://user-images.githubusercontent.com/157042/30103172-8568b1f8-92fa-11e7-949c-b1ae0beef528.png">

Tax amount is rounded and it makes the total sum to be 10610,16 + 1909,83 - 12520 = **-0,01**

**Comment by Andrii-Deineha** [comment]
Hi @dankocherga, Thanks for reporting this issue. Unfortunately, I could not reproduce the issue as you described it. 
![tax rate 100 discount results](https://user-images.githubusercontent.com/27814119/30324497-ee680ad6-97c9-11e7-808a-ad15d4958708.jpg)

**Comment by magento-engcom-team** [comment]
@dankocherga Please provide the detailed steps we must follow to reproduce this issue. Screenshots or logs would be helpful, too.

**Comment by dankocherga** [comment]
@magento-engcom-team 
I have created a Magento database back-up with products and tax rates prepared: https://www.dropbox.com/s/i5kxzqg671zb9nh/1505994954_db_demo.sql.sql.zip?dl=0

**Screenshot of the issue in the cart:**
<img width="1307" alt="2017-09-21 14 57 37" src="https://user-images.githubusercontent.com/157042/30694567-3b96ccb0-9edd-11e7-8826-5c3e72967bd3.png">

**Screenshots of my tax settings:**
<img width="644" alt="2017-09-21 14 58 40" src="https://user-images.githubusercontent.com/157042/30694620-728ce236-9edd-11e7-8055-279cc6ff4199.png">

<img width="654" alt="2017-09-21 14 58 47" src="https://user-images.githubusercontent.com/157042/30694625-7a08cc0a-9edd-11e7-96ab-c611ba9489df.png">

<img width="644" alt="2017-09-21 14 58 53" src="https://user-images.githubusercontent.com/157042/30694631-7edec928-9edd-11e7-93a4-35e9ef610224.png">

**Comment by okorshenko** [comment]
@dankocherga thank you for update

**Comment by magento-engcom-team** [comment]
@dankocherga, thank you for your report.
We've created internal ticket(s) MAGETWO-77433 to track progress on the issue.

**Comment by jonashrem** [comment]
I'm working on it #SQUASHTOBERFEST

**Comment by jonashrem** [comment]
I see this is quite more complicated, than I initially assumed. So I would like to let this handle someone more experienced with magento2 code.

Please reassign this one.

**Comment by jonashrem** [comment]
@okorshenko  please remove my assignment here.

Sorry for the trouble.

**Comment by ilnytskyi** [comment]
So it is really related to rounding issue but we cannot just change the precision.
According to current conditions we can see that: 
`"subtotal_incl_tax": 12519.99,` and  `"discount_amount": -12520,` give negative result.
If we take a look at items we can see 
`"discount_percent": 100,`
and problem is that in **Product B**
`"discount_amount": 6000,` is bigger than `"row_total_incl_tax": 5999.99,`
I propose to understand this issue as _Incorrect Discount Calculation_ since we already can't create a price rule with a discount bigger than **100%** we should not be able to have negative prices.

I propose to introduce additional validation for discount. e.g :
It should not be bigger than _row_total_incl_tax_ or so, and in case of **FALSE** we can set _discount_amount = row_total_incl_tax_

``` 
        "price_incl_tax": 3000,
        "base_price_incl_tax": 3000,
        "row_total_incl_tax": 5999.99,
        "base_row_total_incl_tax": 5999.99
```
btw here we can see incorrect _price_incl_tax_ but correct _row_total_incl_tax_

```{
  "payment_methods": [
    {
      "code": "cashondelivery",
      "title": "Cash On Delivery"
    },
    {
      "code": "purchaseorder",
      "title": "Purchase Order"
    }
  ],
  "totals": {
    "grand_total": 0,
    "base_grand_total": -0.01,
    "subtotal": 10610.16,
    "base_subtotal": 10610.16,
    "discount_amount": -12520,
    "base_discount_amount": -12520,
    "subtotal_with_discount": -1909.84,
    "base_subtotal_with_discount": -1909.84,
    "shipping_amount": 0,
    "base_shipping_amount": 0,
    "shipping_discount_amount": 0,
    "base_shipping_discount_amount": 0,
    "tax_amount": 1909.83,
    "base_tax_amount": 1909.83,
    "weee_tax_applied_amount": null,
    "shipping_tax_amount": 0,
    "base_shipping_tax_amount": 0,
    "subtotal_incl_tax": 12519.99,
    "shipping_incl_tax": 0,
    "base_shipping_incl_tax": 0,
    "base_currency_code": "USD",
    "quote_currency_code": "USD",
    "coupon_code": "FREE",
    "items_qty": 4,
    "items": [
      {
        "item_id": 15,
        "price": 2762.71,
        "base_price": 2762.71,
        "qty": 2,
        "row_total": 5525.42,
        "base_row_total": 5525.42,
        "row_total_with_discount": 0,
        "tax_amount": 994.58,
        "base_tax_amount": 994.58,
        "tax_percent": 18,
        "discount_amount": 6520,
        "base_discount_amount": 6520,
        "discount_percent": 100,
        "price_incl_tax": 3260,
        "base_price_incl_tax": 3260,
        "row_total_incl_tax": 6520,
        "base_row_total_incl_tax": 6520,
        "options": "[]",
        "weee_tax_applied_amount": null,
        "weee_tax_applied": null,
        "name": "Product A"
      },
      {
        "item_id": 16,
        "price": 2542.37,
        "base_price": 2542.37,
        "qty": 2,
        "row_total": 5084.74,
        "base_row_total": 5084.74,
        "row_total_with_discount": 0,
        "tax_amount": 915.25,
        "base_tax_amount": 915.25,
        "tax_percent": 18,
        "discount_amount": 6000,
        "base_discount_amount": 6000,
        "discount_percent": 100,
        "price_incl_tax": 3000,
        "base_price_incl_tax": 3000,
        "row_total_incl_tax": 5999.99,
        "base_row_total_incl_tax": 5999.99,
        "options": "[]",
        "weee_tax_applied_amount": null,
        "weee_tax_applied": null,
        "name": "Product B"
      }
    ],
    "total_segments": [
      {
        "code": "subtotal",
        "title": "Suma cz\u0119\u015bciowa",
        "value": 12519.99
      },
      {
        "code": "shipping",
        "title": "Wysy\u0142ka i obs\u0142uga",
        "value": 0
      },
      {
        "code": "discount",
        "title": "Zni\u017cka (FREE)",
        "value": -12520
      },
      {
        "code": "tax",
        "title": "Podatek",
        "value": 1909.83,
        "extension_attributes": {
          "tax_grandtotal_details": [
            {
              "amount": 1909.83,
              "rates": [
                {
                  "percent": "18",
                  "title": "test_rule"
                }
              ],
              "group_id": 1
            }
          ]
        }
      },
      {
        "code": "grand_total",
        "title": "\u0141\u0105cznie",
        "value": -0.01,
        "area": "footer"
      }
    ]
  }
}```

**Comment by magento-engcom-team** [comment]
Hi @dankocherga. Thank you for your report. 
The issue has been fixed in magento/magento2#14468 by @ilnytskyi in 2.3-develop branch
Related commit(s):
 - [24163073edde6d7cec9382d5ce994d0d0e3481d5](https://github.com/magento/magento2/commit/24163073edde6d7cec9382d5ce994d0d0e3481d5)
 - [32c51d492176762acd841740d7f415e60bd4ad26](https://github.com/magento/magento2/commit/32c51d492176762acd841740d7f415e60bd4ad26)
 - [de1840fb262e0fc57485c8d4db3be97d89faa68f](https://github.com/magento/magento2/commit/de1840fb262e0fc57485c8d4db3be97d89faa68f)
 - [71db984187a31aa9f194177d2764ac30c3946202](https://github.com/magento/magento2/commit/71db984187a31aa9f194177d2764ac30c3946202)
 - [35ec386d9c28a46ab97168ba5f037d9e40c59ead](https://github.com/magento/magento2/commit/35ec386d9c28a46ab97168ba5f037d9e40c59ead)
 - [93e44f8cb1a9de26c7587d97fa4abb7ae6379725](https://github.com/magento/magento2/commit/93e44f8cb1a9de26c7587d97fa4abb7ae6379725)
 - [f3069957250df9acc65392288df823f13e7cf351](https://github.com/magento/magento2/commit/f3069957250df9acc65392288df823f13e7cf351)
 - [317fe620f7e4858d1ea76025e46edc0ed175e7d9](https://github.com/magento/magento2/commit/317fe620f7e4858d1ea76025e46edc0ed175e7d9)
 - [3eed731c22f99b18a161de68e5d4a434d9b57cf7](https://github.com/magento/magento2/commit/3eed731c22f99b18a161de68e5d4a434d9b57cf7)
 - [801b428172a222599320e6a3a449832247265b49](https://github.com/magento/magento2/commit/801b428172a222599320e6a3a449832247265b49)
 - [1610ed3d6e0ae24b794fa4a728e6ea6897b54992](https://github.com/magento/magento2/commit/1610ed3d6e0ae24b794fa4a728e6ea6897b54992)

The fix will be available with the upcoming 2.3.0 release.

**Comment by magento-engcom-team** [comment]
Hi @TomashKhamlai. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if your want to validate it one more time, please, go though the following instruction:
- [x] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [x] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by Ctucker9233** [comment]
Backport to 2.2 created #22188

**Comment by magento-engcom-team** [comment]
Hi @dankocherga. Thank you for your report. 
The issue has been fixed in magento/magento2#22227 by @ilnytskyi in 2.2-develop branch
Related commit(s):
 - [f68cff7feeacdd98faf7365fa676f9e867bffce3](https://github.com/magento/magento2/commit/f68cff7feeacdd98faf7365fa676f9e867bffce3)
 - [8501089751d909786546e2d1631c7f6647d3cded](https://github.com/magento/magento2/commit/8501089751d909786546e2d1631c7f6647d3cded)
 - [7bbc018f04dfc866c89b3816d2b277e1a940d884](https://github.com/magento/magento2/commit/7bbc018f04dfc866c89b3816d2b277e1a940d884)
 - [b30b1bb981e7168b279fb19454072379e96a8961](https://github.com/magento/magento2/commit/b30b1bb981e7168b279fb19454072379e96a8961)
 - [a7f65ac40bcecab88a108f40e3ecd0000a42c6bb](https://github.com/magento/magento2/commit/a7f65ac40bcecab88a108f40e3ecd0000a42c6bb)
 - [0d1a0b0de214004d60bfe564f00f3e9646916f45](https://github.com/magento/magento2/commit/0d1a0b0de214004d60bfe564f00f3e9646916f45)
 - [ea7993768b88465a71c8acba5e84923a4abfb546](https://github.com/magento/magento2/commit/ea7993768b88465a71c8acba5e84923a4abfb546)
 - [ec8bbce2059ace80602b1583f9e9783b2ad6a219](https://github.com/magento/magento2/commit/ec8bbce2059ace80602b1583f9e9783b2ad6a219)
 - [17f7b776c52f6268c8578962985ceb9d1d358c3a](https://github.com/magento/magento2/commit/17f7b776c52f6268c8578962985ceb9d1d358c3a)
 - [d5be0aa37df4dcace02c7a6275cce616062231c4](https://github.com/magento/magento2/commit/d5be0aa37df4dcace02c7a6275cce616062231c4)
 - [98a1561c4b0fbe2cc09000056fa82daa1c39964d](https://github.com/magento/magento2/commit/98a1561c4b0fbe2cc09000056fa82daa1c39964d)

The fix will be available with the upcoming 2.2.9 release.
